src/udp/socket.rs

```diff
@@ -518,264 +535,234 @@ quick_error! {
-                        Ok(HolePunchMsg::Ack) => send_ack_ack_and_proceed(ctx, socket),
-                        Ok(HolePunchMsg::AckAck) => send_final_ack_acks(ctx, socket),
-                        Ok(HolePunchMsg::Choose) => future::ok((socket, true)).into_boxed(),
-                    }
-                }
+                        HolePunchMsg::Syn
+                    },
+                    HolePunchingPhase::Ack => {
+                        HolePunchMsg::Ack
+                    },
+                    HolePunchingPhase::AckAck {
+                        ref mut ack_acks_sent,
+                        received_ack_ack,
+                    } => {
+                        if *ack_acks_sent >= 5 && received_ack_ack {
+                            return Ok(Async::Ready((unwrap!(self.socket.take()), false)));
+                        }
+                        *ack_acks_sent += 1;
+                        HolePunchMsg::AckAck
+                    },
+                };
+                self.send_msg(&msg);
+                continue;
             }
-        })
-        .into_boxed()
-}
-
-// send an ack-ack to the remote peer, then finish hole punching.
-fn send_ack_ack_and_proceed(
-    mut ctx: HolePunching,
-    socket: WithAddress,
-) -> BoxFuture<(WithAddress, bool), HolePunchError> {
-    trace!(
-        "sending ack-ack #{} to {}",
-        ctx.ack_acks_sent,
-        socket.remote_addr()
-    );
-    ctx.send_msg(&HolePunchMsg::AckAck, socket)
-        .and_then(move |socket| {
-            ctx.ack_acks_sent += 1;
-            recv_from_ack_ack(ctx, socket)
-        })
-        .into_boxed()
-}
 
-// we have just sent an ack-ack to the remote peer, listen for incoming packets until it's time to
-// send another one.
-fn recv_from_ack_ack(
-    ctx: HolePunching,
-    socket: WithAddress,
-) -> BoxFuture<(WithAddress, bool), HolePunchError> {
-    socket
-        .with_timeout_at(ctx.msg_timeout(), &ctx.handle)
-        .into_future()
-        .map_err(|(e, _)| HolePunchError::ReadMessage(e))
-        .and_then(move |(msg_opt, socket_timeout)| {
-            let socket = socket_timeout.into_inner();
-            match msg_opt {
-                // TODO: broooooken
-                None => send_ack_ack_and_proceed(ctx, socket),
-                Some(recv_msg) => {
-                    match ctx.decrypt_msg(&recv_msg) {
-                        Err(e) => {
-                            warn!("error deserializing packet from peer: {:?}", e);
-                            recv_from_ack_ack(ctx, socket)
-                        }
-                        Ok(HolePunchMsg::Syn) => recv_from_ack_ack(ctx, socket),
-                        Ok(HolePunchMsg::Ack) => send_ack_ack_and_proceed(ctx, socket),
-                        Ok(HolePunchMsg::AckAck) => send_final_ack_acks(ctx, socket),
-                        Ok(HolePunchMsg::Choose) => future::ok((socket, true)).into_boxed(),
+            match self.recv_msg()? {
```

Comment:
This is a lot of branching in one function, I guess https://en.wikipedia.org/wiki/Cyclomatic_complexity is quite big in here. Why not split this hole message matching into separate function and then split each phase handling into separate functions as well?